src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -1040,16 +1041,18 @@ private void EnumerateChars(Action<int, char[], int> action)
             s_charArrayPool.Free(buffer);
         }
 
-        private int[] ParseLineStarts()
+        private SegmentedList<int> ParseLineStarts()
         {
             // Corner case check
             if (0 == this.Length)
             {
-                return new[] { 0 };
+                return [0];
             }
 
-            var lineStarts = ArrayBuilder<int>.GetInstance();
-            lineStarts.Add(0); // there is always the first line
+            var lineStarts = new SegmentedList<int>()
```

Comment:
nit: consider changing var to SegmentedList<int> here and using a collection-expr